corehq/motech/openmrs/repeater_helpers.py

```diff
@@ -622,6 +640,67 @@ def create_patient(requests, info, case_config):
             return get_patient_by_uuid(requests, response.json()['uuid'])
 
 
+@quickcache(['requests.domain_name', 'requests.base_url', 'identifier_type'])
+def get_identifier_source_id(requests, identifier_type):
+    """
+    Returns the ID of the identifier source to be used for generating
+    values for identifiers of the given type.
+    """
+    # The idgen module doesn't offer an API to list identifier sources.
+    # Log in using a requests Session, and scrape /module/idgen/manageIdentifierSources.list
+    # (Urgh.)
+
+    response = requests.get('/ws/rest/v1/patientidentifiertype/{}'.format(identifier_type))
+    identifier_type_name = response.json()['name']
+
+    session = Session()
+    login_data = {
+        'uname': requests.username,
+        'pw': requests.password,
+        'submit': 'Log In',
+        'redirect': '',
+        'refererURL': '',
+    }
+    session.post(
+        requests.get_url('/ms/legacyui/loginServlet'),
+        login_data,
+        verify=requests.verify
+    )
+
+    response = session.get(
+        requests.get_url('/module/idgen/manageIdentifierSources.list'),
+        verify=requests.verify
+    )
+    tree = html.fromstring(response.content)
+    for row in tree.xpath('//table[@id="sourceTable"]/tbody/tr'):
+        ident_type, source_type, source_name, actions = row.xpath('td')
+        if ident_type.text == identifier_type_name:
+            onclick = actions.xpath('button')[1].attrib['onclick']
+            match = re.match(r"document\.location\.href='viewIdentifierSource\.form\?source=(\d+)';", onclick)
+            source_id = match.group(1)
+            return source_id
+
+
+def generate_identifier(requests, identifier_type):
+    """
+    Calls the idgen module's generateIdentifier endpoint
+
+    Identifier source ID is determined from `identifier_type`. If
+    `identifier_type` doesn't have an identifier source, return None.
+    If the identifier source doesn't return an identifier, return None.
+    If anything goes wrong ... return None.
+    """
+    try:
+        source_id = get_identifier_source_id(requests, identifier_type)
+        if source_id:
+            # Example request: http://www.example.com/openmrs/module/idgen/generateIdentifier.form?source=1
+            response = requests.get('/module/idgen/generateIdentifier.form', {'source': source_id})
+            # Example response: {"identifiers": ["CHR203007"]}
+            return response.json()['identifiers'][0]
+    except:
```

Comment:
There's a lot of stuff going on in here - are there any errors you expect to see regularly, like 404s?  If you really can't let other exceptions bubble up here, I'd recommend adding a `notify_exception` (or `soft_assert`) for all errors besides a handful you anticipate and know to be acceptable.  For instance, if the page gets a redesign and the html parsing breaks, we'd have no way of knowing that's where the problem is.

Reply:
Good call!